### PR TITLE
Some datastructures and A* tweaks

### PR DESCRIPTION
**Pull request by Menshin**
### Some datastructures and A\* tweaks
- Made the Priority Queue insertion use dichotomic search to find the position of the inserted element
- Added a generic (array based) heap data structure
- Changed A\* to use a heap for the open set instead of a priority queue
- Fixed A\* not checking for already processed turfs (shame on whoever fiddled with A*)
- Fixed A\* number of traversed nodes not being updated on new route
- Changed the heuristic proc for every A\* calls to /turf/proc/Distance_cardinal, since A\* now only progress in cardinal directions

Apparently A\* had some changes in the recent past, as it:
- now only progress in cardinal directions,
- <s> only progress on simulated turfs </s>,
-  <s>check for adjacent turf access if an id is specified (mobs, robots) </s>,
-  <s>only check adjacent turfs with atmos if no id is specified (meaning everything without an id is considered to spread like a disease) </s>.

The old procs that were allowing A\* progression through space or with diagonal movements are gone, as far as i can tell.

### CODE REVIEW

**Comment by GunHog** [comment]
Please add the ability to path through space with an appropriate argument passed. This will fix Floorbots and certain Changeling stings.

File: code/__DATASTRUCTURES/heap.dm [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	if(!L.len)
+		return 0
+	. = L[1]
+
+	L[1] = L[L.len]
+	L.Cut(L.len)
+
+	Sink(1)
+
+	return .
```
Comment (RemieRichards) [inline code/__DATASTRUCTURES/heap.dm]: remove the ```return``` here, ```return .``` is implied, and forcing it explicitly by writing ```return``` is actually slower.

Comment (Menshin) [inline code/__DATASTRUCTURES/heap.dm]: Byond never ceases to amaze me.

File: code/__DATASTRUCTURES/heap.dm [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+//Replaces a given node so it verify the heap condition
+/Heap/proc/ReSort(atom/A)
+	var/index = L.Find(A)
+
+	Swim(index)
+	Sink(index)
+
+/Heap/proc/List()
+	var/list/ret = L.Copy()
```
Comment (RemieRichards) [inline code/__DATASTRUCTURES/heap.dm]: ```. = L.Copy()``` without the ```return```, same as above, will be faster.

**Comment by RemieRichards** [comment]
:+1: more data structures, MORE DATA

File: code/__DATASTRUCTURES/heap.dm [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	. = L[1]
+
+	L[1] = L[L.len]
+	L.Cut(L.len)
+
+	Sink(1)
+
+//Get a node up to its right position in the heap
+/Heap/proc/Swim(var/index)
+	var/parent = round(index / 2)
```
Comment (RemieRichards) [inline code/__DATASTRUCTURES/heap.dm]: ```*0.5``` is faster than ```/2```
As these are datastructures, extra speed is nice to have.

**Comment by optimumtact** [comment]
:+1: good changes here thanks for the fixes to A*

**Comment by GunHog** [comment]
Please do not forget about the space travel thing.

**Comment by Menshin** [comment]
#### Continuing

* (Re)Made A* adjacent turfs checking proc generic
* Wrote a ```/reachableAdjacentAtmosTurfs``` proc that checks if a turf can be accessed via atmos
* Made airborne disease spreading only via atmos
* A* old turfs checking proc can now take into account unsimulated turfs using an argument
* Fixes ling stings not working in space (fixes #7560) and floorbot not repairing hull breaches (fixes #10362)

**Comment by optimumtact** [comment]
>Wrote a /reachableAdjacentAtmosTurfs proc that checks if a turf can be accessed via atmos

I thought we had one of these already?

**Comment by Menshin** [comment]
You're probably thinking of ```CanAtmosPass()``` that the proc ```/reachableAdjacentAtmosTurfs``` uses to determine what adjacent turfs have to be treated by A*, when used as the ```adjacent``` proc.

**Comment by Menshin** [comment]
Finally tweaked the "max length path" logic for A* (no need to try for nodes 30 tiles away from start if the path is supposed to be 5 tiles max length).

Should be ready to merge, unless i missed something

**Pull request merged**